src/Arcus.EventGrid/Contracts/Event.cs

```diff
@@ -108,30 +168,78 @@ public string DataVersion
         {
             get
             {
-                if (_eventGridEvent is null)
+                LoadRawInput();
+
+                if (_isEventGridEvent)
                 {
-                    throw new InvalidOperationException(
-                        "Cannot get the data version of this event because it represents a Cloud Event; which don't have any schema version of the data object information");
+                    return _eventGridEvent.DataVersion;
                 }
 
-                return _eventGridEvent.DataVersion;
+                throw new InvalidOperationException(
+                    "Cannot get the data version of this event because it represents a Cloud Event; which don't have any schema version of the data object information");
+
             }
         }
 
         /// <summary>
         ///     The time the event is generated based on the provider's UTC time.
         /// </summary>
-        public DateTimeOffset EventTime => _eventGridEvent?.EventTime ?? _cloudEvent?.Time ?? default(DateTimeOffset);
+        public DateTimeOffset EventTime
+        {
+            get
+            {
+                LoadRawInput();
+
+                if (_isCloudEvent)
```

Comment:
All of this magic should be done when we initialize the object, after that the object should be pretty straight forward.

The things that you are doing here, should've already been done in the parser.

Parser should:
1. Determine inbound schema (Event Grid vs CloudEvents) and version of the spec
2. Create corresponding parser that knows how to map that to our event which creates an instance of this object
3. This object should not know anything about Event Grid or CloudEvents

Reply:
I updated the this in respect to your comment.

Reply:
Resolving as I updated this to that.